### PR TITLE
Refactoring based on comments from previous PRs.

### PR DESCRIPTION
**Pull request by smcheema**
Addresses feedback from PRs #7 & #8 .

### CODE REVIEW

**Review by musanimu — APPROVED** [review]

File: allocator.go
Comment (irfansharif) [inline allocator.go:148]: What you had before was more right, options only make sense during initialization should be passed in during initialization. Right now it's possible to `a := allocator.New(); a.Allocate(); a.WithMatchingTagConstraint()`/ How about including these variadic arguments as part of `allocator.New()`?

Comment (smcheema) [reply allocator.go:148]: Agreed that `new` would be the easiest, it helps to do this there too instead of creating an `opts` struct initialized with each field equalling `false`.

Comment (smcheema) [reply allocator.go:148]: Pushed the change, @irfansharif please let me know if this is better.

File: allocator.go
Comment (irfansharif) [reply allocator.go:34]: Edit: I guess this works, but it's definitely bad and an anti-pattern: https://stackoverflow.com/questions/37952354/golang-public-method-to-private-struct-is-this-have-any-use-case

These public methods on private types will not show up in your go docs. It'd be worthwhile to start adding comments and seeing how your docs would look like using the `godoc` tool.

Comment (smcheema) [reply allocator.go]: TIL, I made the change to prevent manual instantiation of the `allocator` struct, will revert.

Comment (smcheema) [reply allocator.go]: Re: `godoc`, I have a PR almost done for resource-balancing, I've been adding most of the comments there, since this PR is near completion, I'll hold off on adding those here and instead let them come up inside the next PR I'll be posting.

File: allocator.go
Comment (smcheema) [reply allocator.go]: Minor: `Typo: In word 'unassignable` is what GoLand complains about. Hate warnings, so put this down as `unAssignableNodes` versus `unassignableNodes`.

File: allocator.go
Comment (smcheema) [reply allocator.go:148]: Ah right zero-values, good call, added this into the `New()` method.

**Review by irfansharif — APPROVED** [review]
LGTM, though for commit hygiene it's good to collapse all your commits in the PR if there's been internal iteration. Try `git rebase -i HEAD~3` and changing the verb in your last two commits to `fixup`.

**Comment by smcheema** [comment]
Thanks @irfansharif! Question about the interactive rebasing suggestion, is that to keep the commit history clean once the PR is merged? If so, isn't this done with more ease using `Squash and merge` with the appropriate verb on GitHub? Or is this meant to be help during the lifecycle of the PR?

**Comment by irfansharif** [comment]
Both work, but `rebase` is more general purpose. Do whatever you're more comfortable with.

**Pull request merged**
